vm-virtio/src/iommu.rs

```diff
@@ -696,12 +742,20 @@ impl Iommu {
                 acked_features: 0u64,
                 config,
                 mapping: mapping.clone(),
+                ext_mapping: BTreeMap::new(),
                 queue_evts: None,
                 interrupt_cb: None,
             },
             mapping,
         ))
     }
+
+    pub fn set_external_mapping(
+        &mut self,
+        ext_mapping: BTreeMap<u32, Arc<dyn ExternalDmaMapping>>,
+    ) {
+        self.ext_mapping = ext_mapping;
+    }
```

Comment:
This is a one shot API. In my mind, it has a few drawbacks:
1) It forces the caller to use the same data structure as the virtio-iommu implementation
2) You can call it more than once from e.g. your VFIO and vhost-user setup routines
3) It wont allow you to dynamically expand your list of devices with e.g. device hotplug

My suggestion would be to have a
```Rust
    pub fn add_external_mapping(
        &mut self,
        device_id: u32,
        mapping: Arc<dyn ExternalDmaMapping>,
    ) {
        // Insert mapping
    }
```

Reply:
Yes I can change it into this function, but nothing will ever be dynamic since the list has to be complete once the virtio-iommu `activate()` is being triggered, otherwise the devices will be ignored.
About hotplug, unless the IORT ACPI dependency goes away, we will never be able to hotplug devices to the virtual IOMMU.

Reply:
> Yes I can change it into this function, but nothing will ever be dynamic since the list has to be complete once the virtio-iommu `activate()` is being triggered, otherwise the devices will be ignored.

Oh, I thought you could dynamically attach endpoints to the IOMMU device. Still, you want to be able to attach external devices more than once for supporting e.g. vhost-user devices, right? 

> About hotplug, unless the IORT ACPI dependency goes away, we will never be able to hotplug devices to the virtual IOMMU.

Agreed, although that's more of an independent implementation constraint.

Reply:
Done